### PR TITLE
feat(workflows-sdk,core-flows,medusa,types): add workflow to update promotions to cart

### PR DESCRIPTION
**Pull request by riqwan**
what:

- adds API + workflow to add/remove promotions in a cart
- minor fixes in promotions module
- minor type fixes in cart module
- typing fix in workflows-sdk (Thanks @adrien2p)
- fix step result in workflows-sdk (Thanks @adrien2p)

RESOLVES CORE-1768

### CODE REVIEW

File: packages/workflows-sdk/src/utils/composer/create-step.ts [hunk 1]
```diff
@@ -202,6 +202,8 @@ function applyStep<
         ret.__step__ = newStepName
         WorkflowManager.update(this.workflowId, this.flow, this.handlers)
 
+        ret.__step__ = newStepName
```
Comment (sradevski) [inline packages/workflows-sdk/src/utils/composer/create-step.ts]: Why is this needed again?

File: packages/promotion/src/utils/compute-actions/shipping-methods.ts [hunk 2]
```diff
@@ -95,14 +99,18 @@ export function applyPromotionToShippingMethods(
     const totalApplicableValue = shippingMethods!.reduce((acc, method) => {
       const appliedPromoValue = methodIdPromoValueMap.get(method.id) || 0
 
-      return acc + method.subtotal - appliedPromoValue
+      return acc + (method.subtotal || 0) - appliedPromoValue
```
Comment (sradevski) [inline packages/promotion/src/utils/compute-actions/shipping-methods.ts]: I know it doesn't matter in this case, but it's nice practice to rely on `??` instead :)

File: packages/promotion/src/utils/compute-actions/items.ts [hunk 3]
```diff
@@ -112,14 +117,18 @@ export function applyPromotionToItems(
...
-      return (
-        acc +
-        (method.subtotal / method.quantity) * method.quantity -
-        appliedPromoValue
-      )
+      const perItemCost = method.subtotal
+        ? method.subtotal / method.quantity
+        : 0
+
+      return acc + perItemCost * method.quantity - appliedPromoValue
```
Comment (sradevski) [inline packages/promotion/src/utils/compute-actions/items.ts:124]: This is one of the scenarios where you'd want to do some rounding (or maybe higher-up at the caller of this function, but we'll need to round somewhere)

File: integration-tests/plugins/__tests__/cart/store/add-promotions-to-cart.spec.ts [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+              operator: "eq",
+              values: "prod_mat",
+            },
+          ],
+        },
+      })
+
+      const cart = await cartModuleService.create({
+        currency_code: "usd",
+        items: [
```
Comment (sradevski) [inline integration-tests/plugins/__tests__/cart/store/add-promotions-to-cart.spec.ts:91]: Maybe we can create some fixtures that you can reuse, esp. in tests where the creation is just a matter of preparing for the actual test.

File: packages/workflows-sdk/src/utils/composer/create-step.ts (hunk 1 above)
Comment (adrien2p) [reply packages/workflows-sdk/src/utils/composer/create-step.ts]: this how we identified the relation between the steps and the data, where they are coming from etc, here with a custom name the step name wasnt replaced leading to issues. In a workflow composer, even though through the typing we tried to give the idea that you are manipulating the data++ in fact they are only object representation of the flow and relations between the steps

Comment (riqwan) [reply packages/workflows-sdk/src/utils/composer/create-step.ts]: @adrien2p seems like you already added it in another PR line:202

File: packages/promotion/src/utils/compute-actions/items.ts (hunk 3 above)
Comment (riqwan) [reply packages/promotion/src/utils/compute-actions/items.ts:124]: Correct, we will be removing this from the promotion module entirely once we have a totals utility, which should take care of the precision and rounding issues more generally.

File: packages/workflows-sdk/src/utils/composer/create-step.ts (hunk 1 above)
Comment (riqwan) [reply packages/workflows-sdk/src/utils/composer/create-step.ts]: removed this, good catch!

File: integration-tests/plugins/__tests__/cart/store/add-promotions-to-cart.spec.ts [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+              adjustments: expect.arrayContaining([
+                expect.objectContaining({
+                  promotion_id: createdPromotion.id,
+                  code: createdPromotion.code,
+                  amount: 1000,
+                }),
+              ]),
+            }),
+            expect.objectContaining({
+              id: expect.not.stringContaining(lineItemAdjustment.id),
```
Comment (olivermrbl) [inline integration-tests/plugins/__tests__/cart/store/add-promotions-to-cart.spec.ts]: question: Isn't this the wrong id to assert on? You are testing for `items` here but use a line item adjustment ID, so I think this is a false positive no?

File: packages/core-flows/src/definition/cart/steps/create-line-item-adjustments.ts [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+  async (createdLineItemAdjustments, { container }) => {
+    const cartModuleService: ICartModuleService = container.resolve(
+      ModuleRegistrationName.CART
+    )
+
+    if (!createdLineItemAdjustments?.length) {
+      return
+    }
+
+    await cartModuleService.removeLineItemAdjustments(
```
Comment (olivermrbl) [inline packages/core-flows/src/definition/cart/steps/create-line-item-adjustments.ts]: comment: Will be changed to `softDeleteLineItemAdjustments` in #6491

File: packages/core-flows/src/definition/cart/steps/create-shipping-method-adjustments.ts [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+  async (createdShippingMethodAdjustments, { container }) => {
+    const cartModuleService: ICartModuleService = container.resolve(
+      ModuleRegistrationName.CART
+    )
+
+    if (!createdShippingMethodAdjustments?.length) {
+      return
+    }
+
+    await cartModuleService.removeShippingMethodAdjustments(
```
Comment (olivermrbl) [inline packages/core-flows/src/definition/cart/steps/create-shipping-method-adjustments.ts]: Same as other comment.

File: packages/core-flows/src/definition/cart/steps/get-actions-to-compute-from-promotions.ts [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+  cart: CartDTO
+  promoCodes: string[]
+  removePromotions: boolean
+}
+
+export const getActionsToComputeFromPromotionsStepId =
+  "get-actions-to-compute-from-promotions"
+export const getActionsToComputeFromPromotionsStep = createStep(
+  getActionsToComputeFromPromotionsStepId,
+  async (data: StepInput, { container }) => {
```
Comment (olivermrbl) [inline packages/core-flows/src/definition/cart/steps/get-actions-to-compute-from-promotions.ts:16]: thought(non-blocking): Would it make sense for this step to be even more generic i.e. just accept `codes` and return the computed actions? All the data extraction can be performed in transformers. Wdyt?

File: packages/core-flows/src/definition/cart/steps/retrieve-cart.ts [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+    const cartModuleService = container.resolve<ICartModuleService>(
+      ModuleRegistrationName.CART
+    )
+
+    const cart = await cartModuleService.retrieve(data.cartId, {
+      relations: [
+        "items",
+        "items.adjustments",
+        "shipping_methods",
+        "shipping_methods.adjustments",
```
Comment (olivermrbl) [inline packages/core-flows/src/definition/cart/steps/retrieve-cart.ts]: todo: Remove in favor of `...data.config`

File: packages/core-flows/src/definition/cart/steps/retrieve-cart.ts [hunk 10]
```diff
@@ -0,0 +1,43 @@
+import { ModuleRegistrationName } from "@medusajs/modules-sdk"
+import { ICartModuleService } from "@medusajs/types"
+import { StepResponse, createStep } from "@medusajs/workflows-sdk"
+
+interface StepInput {
+  cartId: string
```
Comment (olivermrbl) [inline packages/core-flows/src/definition/cart/steps/retrieve-cart.ts:6]: todo: Add config to increase reusability of the step
```suggestion
  cartId: string
  config: FindConfig<CartDTO>
```

File: packages/core-flows/src/definition/cart/workflows/update-cart-promotions.ts [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+  promoCodes: string[]
+  cartId: string
+  removePromotions?: boolean
+}
+
+export const updateCartPromotionsWorkflowId = "update-cart-promotions"
+export const updateCartPromotionsWorkflow = createWorkflow(
+  updateCartPromotionsWorkflowId,
+  (input: WorkflowData<WorkflowInput>): WorkflowData<CartDTO> => {
+    const cart = retrieveCartStep(input)
```
Comment (olivermrbl) [inline packages/core-flows/src/definition/cart/workflows/update-cart-promotions.ts]: todo: Use config
```suggestion
    const cart = retrieveCartStep(input, { relations: [
       "items",
        "items.adjustments",
        "shipping_methods",
        "shipping_methods.adjustments",
    ] })
```

File: packages/medusa/src/api-v2/store/carts/[id]/promotions/route.ts [hunk 12]
```diff
@@ -0,0 +1,42 @@
+import { updateCartPromotionsWorkflow } from "@medusajs/core-flows"
+import { MedusaRequest, MedusaResponse } from "../../../../../types/routing"
+import { StorePostCartsCartPromotionsReq } from "../../validators"
+
+export const POST = async (req: MedusaRequest, res: MedusaResponse) => {
+  const workflow = updateCartPromotionsWorkflow(req.scope)
+  const requestParams = req.validatedBody as StorePostCartsCartPromotionsReq
```
Comment (olivermrbl) [inline packages/medusa/src/api-v2/store/carts/[id]/promotions/route.ts]: todo: Rename
```suggestion
  const payload = req.validatedBody as StorePostCartsCartPromotionsReq
```

File: packages/medusa/src/api-v2/store/carts/[id]/promotions/route.ts [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+
+  const { result, errors } = await workflow.run({
+    input: {
+      promoCodes: requestParams.promo_codes,
```
Comment (olivermrbl) [inline packages/medusa/src/api-v2/store/carts/[id]/promotions/route.ts]:
```suggestion
      promoCodes: payload.promo_codes,
```

File: packages/medusa/src/api-v2/store/carts/[id]/promotions/route.ts [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+  if (Array.isArray(errors) && errors[0]) {
+    throw errors[0].error
+  }
+
+  res.status(200).json({ cart: result })
+}
+
+export const DELETE = async (req: MedusaRequest, res: MedusaResponse) => {
+  const workflow = updateCartPromotionsWorkflow(req.scope)
+  const requestParams = req.validatedBody as StorePostCartsCartPromotionsReq
```
Comment (olivermrbl) [inline packages/medusa/src/api-v2/store/carts/[id]/promotions/route.ts]: todo: Rename
```suggestion
  const payload = req.validatedBody as StorePostCartsCartPromotionsReq
```

File: packages/medusa/src/api-v2/store/carts/[id]/promotions/route.ts [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+
+  const { result, errors } = await workflow.run({
+    input: {
+      promoCodes: requestParams.promo_codes,
```
Comment (olivermrbl) [inline packages/medusa/src/api-v2/store/carts/[id]/promotions/route.ts]:
```suggestion
      promoCodes: payload.promo_codes,
```

File: packages/medusa/src/api-v2/store/carts/middlewares.ts [hunk 16]
```diff
@@ -38,4 +39,9 @@ export const storeCartRoutesMiddlewares: MiddlewareRoute[] = [
     matcher: "/store/carts/:id",
     middlewares: [transformBody(StorePostCartsCartReq)],
   },
+  {
+    method: ["POST", "DELETE"],
+    matcher: "/store/carts/:id/promotions",
+    middlewares: [transformBody(StorePostCartsCartPromotionsReq)],
```
Comment (olivermrbl) [inline packages/medusa/src/api-v2/store/carts/middlewares.ts:46]: suggestion(non-blocking): Create separate type for `DELETE`

**Review by olivermrbl — CHANGES_REQUESTED** [review]
Nice work. Left a couple of comments and todos.

File: integration-tests/plugins/__tests__/cart/store/add-promotions-to-cart.spec.ts (hunk 5 above)
Comment (riqwan) [reply integration-tests/plugins/__tests__/cart/store/add-promotions-to-cart.spec.ts]: ah yes, good catch. This should go under adjustments

File: packages/core-flows/src/definition/cart/steps/get-actions-to-compute-from-promotions.ts (hunk 8 above)
Comment (riqwan) [reply packages/core-flows/src/definition/cart/steps/get-actions-to-compute-from-promotions.ts:16]: Can you elaborate on this one? I don't fully understand.

For us to be able to generate actions to compute, we need the full cart along with the promo codes.

we also need a third parameter - `action`, which can be:

- `add`: adds new promo codes along with already applied promo codes in the cart (case: add promo code to cart)
- `remove`: removes only one promo code from the cart, keeps all existing promo codes (case: remove promo code from cart)
- `replace`: replace the given promo codes, i.e. removes all existing ones and add only the ones in this given promo code array (case: update cart)

The promo codes to pass into the compute actions will vary depending on each of these actions.

File: packages/core-flows/src/definition/cart/steps/retrieve-cart.ts (hunk 10 above)
Comment (riqwan) [reply packages/core-flows/src/definition/cart/steps/retrieve-cart.ts:6]: nice!

File: packages/core-flows/src/definition/cart/steps/get-actions-to-compute-from-promotions.ts (hunk 8 above)
Comment (olivermrbl) [reply packages/core-flows/src/definition/cart/steps/get-actions-to-compute-from-promotions.ts:16]: Yes, I am not sure what I was thinking actually. I take it back :D

**Review by olivermrbl — APPROVED** [review]
LGTM, nice work!
